File: legacy-python/qairu/core/availability.py
"""Вычисление общих свободных окон.

Поверх `intervals.py`. Про Telegram и БД не знает.

Поддерживает:
- повторяющиеся пары (каждую неделю и по чётности недели);
- разовую занятость на дату и на диапазон дат (сессия, поездка);
- буфер на дорогу — расширение каждой занятости на N минут с двух сторон;
- кворум: окна, где свободны не все, а хотя бы K человек.
"""

from __future__ import annotations

from collections.abc import Callable
from dataclasses import dataclass, field
from datetime import date, timedelta

from .intervals import Interval, coverage_windows, invert, merge, normalize, who_is_free

# Чётность недели: 0 — числитель (верхняя), 1 — знаменатель (нижняя)
ODD, EVEN = 0, 1


def parity_from_semester_start(semester_start: date) -> Callable[[date], int]:
    """Чётность недели относительно начала семестра (первая неделя — числитель)."""
    monday = semester_start - timedelta(days=semester_start.weekday())

    def parity_of(day: date) -> int:
        return ((day - monday).days // 7) % 2

    return parity_of


def parity_from_iso_week(day: date) -> int:
    """Запасной вариант: по номеру ISO-недели (нечётная — числитель)."""
    return (day.isocalendar().week + 1) % 2


@dataclass
class PersonSchedule:
    """Занятость одного человека."""

    user_id: int
    name: str
    weekly: dict[int, list[Interval]] = field(default_factory=dict)
    # {чётность: {день недели: интервалы}} — пары через неделю
    weekly_parity: dict[int, dict[int, list[Interval]]] = field(default_factory=dict)
    dated: dict[date, list[Interval]] = field(default_factory=dict)
    # (дата с, дата по, начало, конец) — сессия, поездка, «занят целиком»
    ranges: list[tuple[date, date, int, int]] = field(default_factory=list)
    has_data: bool = True

    def busy_on(self, day: date, parity: int | None = None, buffer_min: int = 0) -> list[Interval]:
        busy: list[Interval] = list(self.weekly.get(day.weekday(), []))
        if parity is not None:
            busy += self.weekly_parity.get(parity, {}).get(day.weekday(), [])
        busy += self.dated.get(day, [])
        for date_from, date_to, start, end in self.ranges:
            if date_from <= day <= date_to:
                busy.append((start, end))
        if buffer_min:
            busy = [(start - buffer_min, end + buffer_min) for start, end in busy]
        return merge(busy)


@dataclass
class Window:
    """Свободное окно и кто в нём свободен."""

    interval: Interval
    free_ids: list[int]

    @property
    def count(self) -> int:
        return len(self.free_ids)


@dataclass
class DayWindows:
    day: date
    windows: list[Window]

    @property
    def has_windows(self) -> bool:
        return bool(self.windows)


@dataclass
class AvailabilityResult:
    days: list[DayWindows]
    participants: list[PersonSchedule]
    missing: list[PersonSchedule]
    day_start: int
    day_end: int
    min_slot: int
    quorum: int
    everyone: bool  # True, если кворум = все участники

    @property
    def any_windows(self) -> bool:
        return any(day.has_windows for day in self.days)


def free_windows_for_day(
    people: list[PersonSchedule],
    day: date,
    day_start: int,
    day_end: int,
    min_slot: int,
    quorum: int | None = None,
    parity_of: Callable[[date], int] | None = None,
    buffer_min: int = 0,
) -> list[Window]:
    """Свободные окна одного дня.

    quorum=None — нужны все; иначе достаточно `quorum` свободных человек.
    """
    if not people:
        return []

    parity = parity_of(day) if parity_of else None
    free_by_user: dict[int, list[Interval]] = {
        person.user_id: invert(
            normalize(person.busy_on(day, parity, buffer_min), day_start, day_end),
            day_start,
            day_end,
        )
        for person in people
    }

    threshold = len(people) if quorum is None else max(1, min(quorum, len(people)))
    windows: list[Window] = []
    for interval, _ in coverage_windows(list(free_by_user.values()), threshold):
        if interval[1] - interval[0] < min_slot:
            continue
        windows.append(Window(interval=interval, free_ids=who_is_free(free_by_user, interval)))
    return windows


def compute_availability(
    people: list[PersonSchedule],
    start_day: date,
    days_ahead: int = 7,
    day_start: int = 8 * 60,
    day_end: int = 22 * 60,
    min_slot: int = 30,
    quorum: int | None = None,
    parity_of: Callable[[date], int] | None = None,
    buffer_min: int = 0,
) -> AvailabilityResult:
    """Свободные окна на `days_ahead` дней вперёд, начиная со `start_day`."""
    with_data = [person for person in people if person.has_data]
    missing = [person for person in people if not person.has_data]

    threshold = len(with_data) if quorum is None else max(1, min(quorum, len(with_data)))

    days = [
        DayWindows(
            day=start_day + timedelta(days=offset),
            windows=free_windows_for_day(
                with_data,
                start_day + timedelta(days=offset),
                day_start,
                day_end,
                min_slot,
                quorum,
                parity_of,
                buffer_min,
            ),
        )
        for offset in range(days_ahead)
    ]

    return AvailabilityResult(
        days=days,
        participants=with_data,
        missing=missing,
        day_start=day_start,
        day_end=day_end,
        min_slot=min_slot,
        quorum=threshold,
        everyone=threshold >= len(with_data),
    )


def top_slots(result: AvailabilityResult, limit: int = 5) -> list[tuple[date, Interval]]:
    """Первые `limit` окон по хронологии — для кнопок выбора времени встречи."""
    slots: list[tuple[date, Interval]] = []
    for day in result.days:
        for window in day.windows:
            slots.append((day.day, window.interval))
            if len(slots) >= limit:
                return slots
    return slots


@dataclass
class HeatCell:
    """Одна клетка сетки: сколько человек свободно на всём её протяжении."""

    start_min: int
    end_min: int
    free_ids: list[int]

    @property
    def count(self) -> int:
        return len(self.free_ids)


@dataclass
class HeatDay:
    day: date
    cells: list[HeatCell]


def heatmap(
    people: list[PersonSchedule],
    start_day: date,
    days_ahead: int = 7,
    day_start: int = 8 * 60,
    day_end: int = 22 * 60,
    step: int = 30,
    parity_of: Callable[[date], int] | None = None,
    buffer_min: int = 0,
) -> list[HeatDay]:
    """Тепловая карта недели: для каждой получасовой клетки — кто свободен.

    Это то, чего не может дать чат: одним взглядом видно, где «почти все»
    свободны, а где провал. Клетка считается свободной только если человек
    свободен на всём её протяжении — половинчатых значений нет намеренно,
    иначе цвет обманывает.
    """
    with_data = [person for person in people if person.has_data]
    result: list[HeatDay] = []

    for offset in range(days_ahead):
        day = start_day + timedelta(days=offset)
        parity = parity_of(day) if parity_of else None
        free_by_user = {
            person.user_id: invert(
                normalize(person.busy_on(day, parity, buffer_min), day_start, day_end),
                day_start,
                day_end,
            )
            for person in with_data
        }
        cells: list[HeatCell] = []
        for start in range(day_start, day_end, step):
            end = min(start + step, day_end)
            cells.append(
                HeatCell(start_min=start, end_min=end,
                         free_ids=who_is_free(free_by_user, (start, end)))
            )
        result.append(HeatDay(day=day, cells=cells))
    return result

File: src/core/availability.ts
/**
 * Вычисление общих свободных окон.
 *
 * Поверх `intervals.ts`. Про Telegram, БД и HTTP не знает.
 *
 * Поддерживает:
 * - повторяющиеся пары (каждую неделю и по чётности недели);
 * - разовую занятость на дату и на диапазон дат (сессия, поездка);
 * - буфер на дорогу — расширение каждой занятости на N минут с двух сторон;
 * - кворум: окна, где свободны не все, а хотя бы K человек.
 */

import { type Interval, coverageWindows, invert, merge, normalize, whoIsFree } from "./intervals";
import { type DateStr, addDays, compareDates, diffDays, weekdayOf } from "./timeutils";

// Чётность недели: 0 — числитель (верхняя), 1 — знаменатель (нижняя)
export const ODD = 0;
export const EVEN = 1;

export type ParityOf = (day: DateStr) => number;

/** Чётность недели относительно начала семестра (первая неделя — числитель). */
export function parityFromSemesterStart(semesterStart: DateStr): ParityOf {
  const monday = addDays(semesterStart, -weekdayOf(semesterStart));
  return (day: DateStr) => {
    const weeks = Math.floor(diffDays(monday, day) / 7);
    return ((weeks % 2) + 2) % 2;
  };
}

/** Запасной вариант: чётность по номеру ISO-недели (нечётная — числитель). */
export function parityFromIsoWeek(day: DateStr): number {
  const [year, month, dayOfMonth] = day.split("-").map(Number);
  const thursday = new Date(Date.UTC(year, month - 1, dayOfMonth));
  // Неделя ISO определяется своим четвергом — он всегда в «правильном» году.
  thursday.setUTCDate(thursday.getUTCDate() - ((thursday.getUTCDay() + 6) % 7) + 3);
  const firstThursday = new Date(Date.UTC(thursday.getUTCFullYear(), 0, 4));
  firstThursday.setUTCDate(
    firstThursday.getUTCDate() - ((firstThursday.getUTCDay() + 6) % 7) + 3,
  );
  const week =
    1 + Math.round((thursday.getTime() - firstThursday.getTime()) / (7 * 86_400_000));
  return (week + 1) % 2;
}

export type DateRangeSlot = {
  dateFrom: DateStr;
  dateTo: DateStr;
  start: number;
  end: number;
};

/** Занятость одного человека. */
export class PersonSchedule {
  userId: number;
  name: string;
  /** день недели -> интервалы (каждую неделю) */
  weekly: Map<number, Interval[]>;
  /** чётность -> день недели -> интервалы (пары через неделю) */
  weeklyParity: Map<number, Map<number, Interval[]>>;
  /** конкретная дата -> интервалы */
  dated: Map<DateStr, Interval[]>;
  /** сессия, поездка, «занят целиком» */
  ranges: DateRangeSlot[];
  hasData: boolean;

  constructor(init: {
    userId: number;
    name: string;
    weekly?: Map<number, Interval[]>;
    weeklyParity?: Map<number, Map<number, Interval[]>>;
    dated?: Map<DateStr, Interval[]>;
    ranges?: DateRangeSlot[];
    hasData?: boolean;
  }) {
    this.userId = init.userId;
    this.name = init.name;
    this.weekly = init.weekly ?? new Map();
    this.weeklyParity = init.weeklyParity ?? new Map();
    this.dated = init.dated ?? new Map();
    this.ranges = init.ranges ?? [];
    this.hasData = init.hasData ?? true;
  }

  busyOn(day: DateStr, parity: number | null = null, bufferMin = 0): Interval[] {
    const weekday = weekdayOf(day);
    let busy: Interval[] = [...(this.weekly.get(weekday) ?? [])];
    if (parity !== null) {
      busy = busy.concat(this.weeklyParity.get(parity)?.get(weekday) ?? []);
    }
    busy = busy.concat(this.dated.get(day) ?? []);
    for (const range of this.ranges) {
      if (compareDates(range.dateFrom, day) <= 0 && compareDates(day, range.dateTo) <= 0) {
        busy.push([range.start, range.end]);
      }
    }
    if (bufferMin) {
      busy = busy.map(([start, end]) => [start - bufferMin, end + bufferMin] as Interval);
    }
    return merge(busy);
  }
}

/** Свободное окно и кто в нём свободен. */
export type Window = {
  interval: Interval;
  freeIds: number[];
};

export type DayWindows = {
  day: DateStr;
  windows: Window[];
};

export type AvailabilityResult = {
  days: DayWindows[];
  participants: PersonSchedule[];
  missing: PersonSchedule[];
  dayStart: number;
  dayEnd: number;
  minSlot: number;
  quorum: number;
  /** true, если кворум = все участники */
  everyone: boolean;
};

export function anyWindows(result: AvailabilityResult): boolean {
  return result.days.some((day) => day.windows.length > 0);
}

function freeByUserFor(
  people: readonly PersonSchedule[],
  day: DateStr,
  dayStart: number,
  dayEnd: number,
  parityOf: ParityOf | null,
  bufferMin: number,
): Map<number, Interval[]> {
  const parity = parityOf ? parityOf(day) : null;
  const map = new Map<number, Interval[]>();
  for (const person of people) {
    map.set(
      person.userId,
      invert(normalize(person.busyOn(day, parity, bufferMin), dayStart, dayEnd), dayStart, dayEnd),
    );
  }
  return map;
}

/**
 * Свободные окна одного дня.
 *
 * quorum=null — нужны все; иначе достаточно `quorum` свободных человек.
 */
export function freeWindowsForDay(
  people: readonly PersonSchedule[],
  day: DateStr,
  dayStart: number,
  dayEnd: number,
  minSlot: number,
  quorum: number | null = null,
  parityOf: ParityOf | null = null,
  bufferMin = 0,
): Window[] {
  if (people.length === 0) return [];

  const freeByUser = freeByUserFor(people, day, dayStart, dayEnd, parityOf, bufferMin);
  const threshold = quorum === null ? people.length : Math.max(1, Math.min(quorum, people.length));

  const windows: Window[] = [];
  for (const [interval] of coverageWindows([...freeByUser.values()], threshold)) {
    if (interval[1] - interval[0] < minSlot) continue;
    windows.push({ interval, freeIds: whoIsFree(freeByUser, interval) });
  }
  return windows;
}

export type ComputeOptions = {
  daysAhead?: number;
  dayStart?: number;
  dayEnd?: number;
  minSlot?: number;
  quorum?: number | null;
  parityOf?: ParityOf | null;
  bufferMin?: number;
};

/** Свободные окна на `daysAhead` дней вперёд, начиная со `startDay`. */
export function computeAvailability(
  people: readonly PersonSchedule[],
  startDay: DateStr,
  options: ComputeOptions = {},
): AvailabilityResult {
  const {
    daysAhead = 7,
    dayStart = 8 * 60,
    dayEnd = 22 * 60,
    minSlot = 30,
    quorum = null,
    parityOf = null,
    bufferMin = 0,
  } = options;

  const withData = people.filter((person) => person.hasData);
  const missing = people.filter((person) => !person.hasData);

  const threshold =
    quorum === null ? withData.length : Math.max(1, Math.min(quorum, withData.length));

  const days: DayWindows[] = [];
  for (let offset = 0; offset < daysAhead; offset += 1) {
    const day = addDays(startDay, offset);
    days.push({
      day,
      windows: freeWindowsForDay(
        withData,
        day,
        dayStart,
        dayEnd,
        minSlot,
        quorum,
        parityOf,
        bufferMin,
      ),
    });
  }

  return {
    days,
    participants: withData,
    missing,
    dayStart,
    dayEnd,
    minSlot,
    quorum: threshold,
    everyone: threshold >= withData.length,
  };
}

/** Первые `limit` окон по хронологии — для кнопок выбора времени встречи. */
export function topSlots(
  result: AvailabilityResult,
  limit = 5,
): { day: DateStr; interval: Interval }[] {
  const slots: { day: DateStr; interval: Interval }[] = [];
  for (const day of result.days) {
    for (const window of day.windows) {
      slots.push({ day: day.day, interval: window.interval });
      if (slots.length >= limit) return slots;
    }
  }
  return slots;
}

/** Вариант встречи ровно заданной длины и кто свободен на всём её протяжении. */
export type FixedSlot = {
  interval: Interval;
  freeIds: number[];
};

export type DaySlots = {
  day: DateStr;
  slots: FixedSlot[];
};

/**
 * Варианты встречи ровно на `duration` минут, с шагом `step`.
 *
 * В отличие от окон (сплошной промежуток «от и до»), здесь каждое начало —
 * отдельная кнопка «Назначить»: 10:00–11:00, 10:30–11:30, … Состав свободных
 * считается честно для каждого варианта отдельно, поэтому при кворуме у
 * 11:00–12:00 может не хватать другого человека, чем у 10:00–11:00.
 *
 * Начала выровнены по сетке `dayStart + k·step`, так варианты совпадают с
 * клетками тепловой карты. Вариант, выходящий за конец дня, не предлагается.
 */
export function slotsOfLength(
  people: readonly PersonSchedule[],
  startDay: DateStr,
  options: ComputeOptions & { duration: number; step?: number },
): { days: DaySlots[]; participants: PersonSchedule[]; quorum: number; everyone: boolean } {
  const {
    daysAhead = 7,
    dayStart = 8 * 60,
    dayEnd = 22 * 60,
    quorum = null,
    parityOf = null,
    bufferMin = 0,
    duration,
    step = 30,
  } = options;

  const withData = people.filter((person) => person.hasData);
  const threshold =
    quorum === null ? withData.length : Math.max(1, Math.min(quorum, withData.length));

  const days: DaySlots[] = [];
  for (let offset = 0; offset < daysAhead; offset += 1) {
    const day = addDays(startDay, offset);
    const slots: FixedSlot[] = [];
    if (withData.length > 0 && duration > 0 && step > 0) {
      const freeByUser = freeByUserFor(withData, day, dayStart, dayEnd, parityOf, bufferMin);
      for (let start = dayStart; start + duration <= dayEnd; start += step) {
        const interval: Interval = [start, start + duration];
        const freeIds = whoIsFree(freeByUser, interval);
        if (freeIds.length >= threshold) slots.push({ interval, freeIds });
      }
    }
    days.push({ day, slots });
  }

  return {
    days,
    participants: withData,
    quorum: threshold,
    everyone: threshold >= withData.length,
  };
}

/** Одна клетка сетки: сколько человек свободно на всём её протяжении. */
export type HeatCell = {
  startMin: number;
  endMin: number;
  freeIds: number[];
};

export type HeatDay = {
  day: DateStr;
  cells: HeatCell[];
};

/**
 * Тепловая карта недели: для каждой получасовой клетки — кто свободен.
 *
 * Это то, чего не может дать чат: одним взглядом видно, где «почти все»
 * свободны, а где провал. Клетка считается свободной только если человек
 * свободен на всём её протяжении — половинчатых значений нет намеренно,
 * иначе цвет обманывает.
 */
export function heatmap(
  people: readonly PersonSchedule[],
  startDay: DateStr,
  options: {
    daysAhead?: number;
    dayStart?: number;
    dayEnd?: number;
    step?: number;
    parityOf?: ParityOf | null;
    bufferMin?: number;
    /** Готовые ряды (например, пары). Без них — отрезки по `step`. */
    rows?: readonly { start: number; end: number }[];
  } = {},
): HeatDay[] {
  const {
    daysAhead = 7,
    dayStart = 8 * 60,
    dayEnd = 22 * 60,
    step = 30,
    parityOf = null,
    bufferMin = 0,
    rows = null,
  } = options;

  const withData = people.filter((person) => person.hasData);
  const result: HeatDay[] = [];

  for (let offset = 0; offset < daysAhead; offset += 1) {
    const day = addDays(startDay, offset);
    const freeByUser = freeByUserFor(withData, day, dayStart, dayEnd, parityOf, bufferMin);
    const cells: HeatCell[] = [];
    if (rows) {
      for (const row of rows) {
        cells.push({ startMin: row.start, endMin: row.end, freeIds: whoIsFree(freeByUser, [row.start, row.end]) });
      }
      result.push({ day, cells });
      continue;
    }
    for (let start = dayStart; start < dayEnd; start += step) {
      const end = Math.min(start + step, dayEnd);
      cells.push({ startMin: start, endMin: end, freeIds: whoIsFree(freeByUser, [start, end]) });
    }
    result.push({ day, cells });
  }
  return result;
}
